cmd/traffic/cmd/manager/service.go

```diff
@@ -268,10 +272,98 @@ func (s *service) Remain(ctx context.Context, req *rpc.RemainRequest) (*empty.Em
 func (s *service) Depart(ctx context.Context, session *rpc.SessionInfo) (*empty.Empty, error) {
 	ctx = managerutil.WithSessionInfo(ctx, session)
 	sessionID := tunnel.SessionID(session.GetSessionId())
-
+	err := s.UpdateLastEngagementTime(ctx, sessionID)
 	// There's no reason for the caller to wait for this removal to complete.
 	go s.state.RemoveSession(context.WithoutCancel(ctx), sessionID)
-	return &empty.Empty{}, nil
+	return &empty.Empty{}, err
+}
+
+func (s *service) removeUnusedAgent(ctx context.Context, sessionID tunnel.SessionID) error {
+	maxIdleTime := managerutil.GetEnv(ctx).AgentMaxIdleTime
+	if maxIdleTime == 0 {
+		// default aka not set is 0, we don't ever remove agents, skip
+		return nil
+	}
+	agentStateYAML := s.configWatcher.GetAgentStateYaml(ctx)
+	var agentStates map[string]AgentState
+	err := yaml.Unmarshal([]byte(agentStateYAML), &agentStates)
+	if err != nil {
+		return fmt.Errorf("err in unmarshalling YAML: %w", err)
+	}
+	agentKey, err := s.getAgentKey(sessionID)
+	if err != nil {
+		return err
+	}
+
+	lastEngagementTime := agentStates[agentKey].LastEngagementTime
+	idleTime := time.Since(lastEngagementTime)
+	if idleTime > maxIdleTime {
+		s.state.RemoveAgentSession(ctx, sessionID)
+	}
+	return nil
+}
+func (s *service) UpdateLastEngagementTime(ctx context.Context, sessionID tunnel.SessionID) error {
```

Comment:
This method must also be called when an engagement starts, and then periodically during the time when the engagement is active, or you risk that an active engagement suddenly gets removed.

Reply:
Yes, that'll work and it's much simpler than what I suggested. If the engagement isn't active, and hasn't been active for some time, then it's OK to remove it. I don't see any flaws in that logic.